horovod/common/ops/cuda/adasum_cuda_kernels.cu

```diff
@@ -0,0 +1,202 @@
+// Copyright 2019 Microsoft. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <iostream>
+#include <stdio.h>
+#include <cuda_fp16.h>
+#include <time.h>
+#include <stdint.h>
+#include <float.h>
+
+#define THREADS_PER_BLOCK 64
+
+__device__ __constant__ double min_guard = DBL_MIN;
+
+template<typename T, typename TACC>
+__global__
+void CudaSingleAdasumKernel(int count, T* a, const T* b, TACC* out) {
+	__shared__ TACC a_coeff;
+	__shared__ TACC b_coeff;
+	int index = threadIdx.x + blockIdx.x * blockDim.x;
+	if (0 == threadIdx.x) {
+		if ((TACC) out[0] >= min_guard) {
+			a_coeff = 1.0 - out[2] / out[0] * 0.5;
+		} else {
+			a_coeff = 1;
+		}
+		if ((TACC) out[1] >= min_guard) {
+			b_coeff = 1.0 - out[2] / out[1] * 0.5;
+		} else {
+			b_coeff = 1;
+		}
+	}
+	__syncthreads();
+	if (count > index){
+		a[index] = (T) ((TACC) a[index] * a_coeff + (TACC) b[index] * b_coeff);
+	}
+}
+
+template<typename T, typename TACC>
+__global__
+void CudaDotProductKernel(int count, const T* a, const T* b, TACC* out) {
```

Comment:
I tried using the cuBLAS operations for this previously, but found that the performance was actually a lot lower with those than writing this directly (~20-25% slower). I don't believe that it's worth the trade off.

Reply:
I'd expect using individual cublas calls to be slower than the fused approach you have, but the benefit is determinism. Is the 20-25% slower end-to-end, or just timing this particular set of operations. @tgaddair do you have any comments on this regarding deterministic operations?

Reply:
In general, my preference is to prefer performance over determinism by default, but provide a means for users to enforce stricter determinism guarantees if desired.  So we could toggle between the fused and unfused ops with an environment variable or `horovodrun` arg, for example.  I feel this would be consistent with the way we handle similar cases (e.g., tensor fusion).  Does that sound reasonable to you all?

Reply:
I think that sounds reasonable. A second deterministic version of this fused kernel approach can be written. Instead of atomic adds, the first kernel should write intermediate block sums to a temporary array. Then a second kernel can be launched to sum the temporary results in the array.

Reply:
I actually had this implemented exactly as you described originally (I think the part below is the leftover portion from me forgetting to remove all that code). The 20-25% slowdown I observed with it was on the pytorch synthetic benchmark example (resnet50 with batch size 32). So it was a significant slowdown in an end-to-end scenario, albeit a very communication bound one. I am not sure how significant of a concern this non determinism is, but I could add back the deterministic path. I mostly removed it because I think the slowdown is too much for any real use case, and it adds a rather large amount of code.